File: Practica_1/src/uiMain/Almacenamiento.java
package uiMain;

import gestorAplicacion.*;

import java.io.Serializable;
import java.sql.Date;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import baseDatos.Deserializador;

public class Almacenamiento implements Serializable {

    // LISTA PARA GUARDAR LOS OBJETOS DE CADA TIPO (SERIALIZACION)

    // lista usurios
    private static List<Usuario> listaUsuarios = new ArrayList<Usuario>();
    // lista empleados
    private static List<Empleado> listaEmpleados = new ArrayList<Empleado>();

    // Lista Habitaciones
    private static List<Lugar> listaHabitaciones = new ArrayList<Lugar>();
    // Lista Lugares para eventos
    private static List<Lugar> listaLugares = new ArrayList<Lugar>();

    // lista reservas
    private static List<Reserva> listaReservas = new ArrayList<Reserva>();
    // lista eventos
    private static List<Evento> listaEventos = new ArrayList<Evento>();
    // lista Servicios
    private static List<Servicio> listaServicios = new ArrayList<Servicio>();

    // lista Facturas
    private static List<Factura> listaFacturas = new ArrayList<Factura>();
    // lista Servicios Externos
    private static List<ServicioExterno> listaServiciosExternos = new ArrayList<ServicioExterno>();

    // lista copiada de main
    private static List<Lugar> habitaciondis = new ArrayList<Lugar>();

    // METODOS DE BUSQUEDA
    public static Usuario buscarUsuario(int id) {
        Iterator<Usuario> iterator = Almacenamiento.getListaUsuarios().iterator();
        while (iterator.hasNext()) {
            Usuario cliente = (Usuario) iterator.next();
            if (cliente.getIdentificacion() == id) {
                return cliente;

            }
        }
        return null;
    }

    public static Empleado buscarEmpleado(int id) {
        Iterator<Empleado> iterator = Almacenamiento.getListaEmpleados().iterator();
        while (iterator.hasNext()) {
            Empleado empleado = (Empleado) iterator.next();
            if (empleado.getIdentificacion() == id) {
                return empleado;

            }
        }
        return null;
    }

    public static Lugar buscarHabitacion(int id) {
        Iterator<Lugar> iterator = Almacenamiento.getListaHabitaciones().iterator();
        while (iterator.hasNext()) {
            Lugar habitacion = (Lugar) iterator.next();
            if (habitacion.getNumero() == id) {
                return habitacion;

            }
        }
        return null;
    }

    public static Lugar buscarLugar(String tipo) {
        Iterator<Lugar> iterator = Almacenamiento.getListaLugares().iterator();
        while (iterator.hasNext()) {
            Lugar lugar = (Lugar) iterator.next();
            if (lugar.getTipo() == tipo) {
                return lugar;

            }
        }
        return null;
    }

    public static Servicio buscarServicio(String nombre) {
        Iterator<Servicio> iterator = Almacenamiento.getListaServicios().iterator();
        while (iterator.hasNext()) {
            Servicio servicio = (Servicio) iterator.next();
            if (servicio.getNombre() == nombre) {
                return servicio;

            }
        }
        return null;
    }

    public static Reserva buscarReserva(int id) {
        Iterator<Reserva> iterator = Almacenamiento.getListaReservas().iterator();
        while (iterator.hasNext()) {
            Reserva reserva = (Reserva) iterator.next();
            if (reserva.getCliente().getIdentificacion() == id) {
                return reserva;

            }
        }
        return null;
    }

    public static Evento buscarEvento(int id) {
        Iterator<Evento> iterator = Almacenamiento.getListaEventos().iterator();
        while (iterator.hasNext()) {
            Evento evento = (Evento) iterator.next();
            if (evento.getCodigo() == id) {
                return evento;

            }
        }
        return null;
    }

    // METODOS PARA CREAR OBJETOS

    public static Usuario crearUsuario(String nombre, int identificacion, int telefono, String tipo,
            String cuentaBancaria,
            ArrayList<String> arrayList) {
        Usuario usuario = new Usuario(nombre, identificacion, telefono, tipo, cuentaBancaria, new ArrayList<Factura>());
        Almacenamiento.listaUsuarios.add(usuario);
        return usuario;
    }

    public static Empleado crearEmpleado(String nombre, int identificacion, int telefono, String cargo) {
        Empleado empleado = new Empleado(nombre, identificacion, cargo);
        Almacenamiento.listaEmpleados.add(empleado);
        return empleado;
    }

    public static Lugar crearHabitacion(int numero, String tipo, int capacidad) {
        Lugar habitacion = new Lugar(numero, tipo, capacidad);
        Almacenamiento.listaHabitaciones.add(habitacion);
        return habitacion;
    }

    public static Lugar crearLugar(String tipo) {
        Lugar lugar = new Lugar(tipo);
        Almacenamiento.listaLugares.add(lugar);
        return lugar;

    }

    public static Servicio crearServico(String nombre, int valor) {
        Servicio servicio = new Servicio(nombre, valor);
        Almacenamiento.listaServicios.add(servicio);
        return servicio;
    }

    public static Reserva crearReserva(String fechaEntrada, String fechaSalida, List<Lugar> habitaciones, float aporte,
            Usuario cliente) {
        Reserva reserva = new Reserva(fechaEntrada, fechaSalida, habitaciones, aporte, cliente);
        Almacenamiento.listaReservas.add(reserva);
        return reserva;
    }

    public static Evento crearEvento(Lugar lugar, Usuario cliente, ArrayList<ServicioExterno> servicios, String fecha,
            int duracion, int numeroAsistentes, ArrayList<Empleado> empleados, int codigo) {
        Evento evento = new Evento(lugar, cliente, servicios, fecha, duracion, numeroAsistentes, empleados, codigo);
        Almacenamiento.listaEventos.add(evento);
        return evento;
    }

    public static Factura crearFactura(Usuario cliente, Empleado empleado, List<Servicio> items, Destinos destino,
			String concepto) {
		Factura factura = new Factura(cliente, empleado, items, destino, concepto);
		listaFacturas.add(factura);
		return factura;
	}

    // METODOS COPIADOS DE LA CLASE MAIN
    // Metodos que tengo que copiar en almacenamiento

    public static String listaUsuarios() {
        Iterator<Usuario> iterator = listaUsuarios.iterator();
        StringBuffer lista = new StringBuffer();
        while (iterator.hasNext()) {
            Usuario usuario = (Usuario) iterator.next();
            lista.append(usuario.informacion() + "\n");
        }
        return lista.toString();
    }

    public static String listaReservas() {
        Iterator<Reserva> iterator = listaReservas.iterator();
        StringBuffer lista = new StringBuffer();
        while (iterator.hasNext()) {
            Reserva reserva = (Reserva) iterator.next();
            lista.append(reserva.toString() + "\n");
        }
        return lista.toString();
    }

    public static String listaFacturascliente(Usuario cliente) {
        Iterator<Factura> iterator = cliente.getListaFacturas().iterator();
        StringBuffer lista = new StringBuffer();
        while (iterator.hasNext()) {
            Factura factura = (Factura) iterator.next();
            lista.append(factura.toString() + "\n");
        }
        return lista.toString();
    }

    public static String listaFacturas() {
        Iterator<Factura> iterator = listaFacturas.iterator();
        StringBuffer lista = new StringBuffer();
        while (iterator.hasNext()) {
            Factura factura = (Factura) iterator.next();
            lista.append(factura.toString() + "\n");
        }
        return lista.toString();
    }

    public static String listaHabitaciones() {
        Iterator<Lugar> iterator = listaHabitaciones.iterator();
        StringBuffer lista = new StringBuffer();
        while (iterator.hasNext()) {
            Lugar habitacion = (Lugar) iterator.next();
            lista.append(habitacion.toString() + "\n");
        }
        return lista.toString();
    }

    // Sobrecarga de metodos
    public static String listaHabitaciones(List<Lugar> esto) {
        Iterator<Lugar> iterator = esto.iterator();
        StringBuffer lista = new StringBuffer();
        while (iterator.hasNext()) {
            Lugar habitacion = (Lugar) iterator.next();
            lista.append(habitacion.toString() + "\n");
        }
        return lista.toString();
    }

    public static void nodisponible(Lugar hb) {
        listaHabitaciones.remove(hb);
    }
/*
    public static String listarDisponibles(String fechaEntrada, String fechaSalida) {

        habitaciondis.clear();
        SimpleDateFormat fecha = new SimpleDateFormat("dd/MM/yyyy");
        Date fEntrada = new Date();
        try {
            fEntrada = fecha.parse(fechaEntrada);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        Date fSalida = new Date();
        try {
            fSalida = fecha.parse(fechaSalida);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        for (Reserva re : listaReservas) {
            for (Lugar h : re.getHabitaciones()) {
                if (fEntrada.after(re.getFechaEntrada()) && fSalida.after(re.getFechaSalida())
                        || fEntrada.before(re.getFechaEntrada()) && fSalida.before(re.getFechaSalida())) {
                    habitaciondis.add(h);
                }
            }
        }
        Iterator<Lugar> iterator = listaHabitaciones.iterator();
        while (iterator.hasNext()) {
            Lugar habitacion = (Lugar) iterator.next();
            habitaciondis.add(habitacion);
        }
        Collections.sort(habitaciondis, new Comparator<Lugar>() {
            public int compare(Lugar p1, Lugar p2) {
                return new Integer(p1.getNumero()).compareTo(new Integer(p2.getNumero()));
            }
        });
        String lista = listaHabitaciones(habitaciondis);
        return lista;

    }
*/
    // CONSTRUCTOR
    public Almacenamiento() {
        Deserializador.deserializar(this);
    }

    // GETERS AND SETTERS
    public static List<Usuario> getListaUsuarios() {
        return listaUsuarios;
    }

    public static void setListaUsuarios(List<Usuario> listaUsuarios) {
        Almacenamiento.listaUsuarios = listaUsuarios;
    }

    public static List<Empleado> getListaEmpleados() {
        return listaEmpleados;
    }

    public static void setListaEmpleados(List<Empleado> listaEmpleados) {
        Almacenamiento.listaEmpleados = listaEmpleados;
    }

    public static List<Lugar> getListaHabitaciones() {
        return listaHabitaciones;
    }

    public static void setListaHabitaciones(List<Lugar> listaHabitaciones) {
        Almacenamiento.listaHabitaciones = listaHabitaciones;
    }

    public static List<Lugar> getListaLugares() {
        return listaLugares;
    }

    public static void setListaLugares(List<Lugar> listaLugares) {
        Almacenamiento.listaLugares = listaLugares;
    }

    public static List<Reserva> getListaReservas() {
        return listaReservas;
    }

    public static void setListaReservas(List<Reserva> listaReservas) {
        Almacenamiento.listaReservas = listaReservas;
    }

    public static List<Evento> getListaEventos() {
        return listaEventos;
    }

    public static void setListaEventos(List<Evento> listaEventos) {
        Almacenamiento.listaEventos = listaEventos;
    }

    public static List<Servicio> getListaServicios() {
        return listaServicios;
    }

    public static void setListaServicios(List<Servicio> listaServicios) {
        Almacenamiento.listaServicios = listaServicios;
    }

    public static List<ServicioExterno> getListaServiciosExternos() {
        return listaServiciosExternos;
    }

    public static void setListaServiciosExternos(List<ServicioExterno> listaServiciosExternos) {
        Almacenamiento.listaServiciosExternos = listaServiciosExternos;
    }

    public static List<Factura> getListaFacturas() {
        return listaFacturas;
    }

    public static void setListaFacturas(List<Factura> listaFacturas) {
        Almacenamiento.listaFacturas = listaFacturas;
    }

}

File: Python/baseDatos/Almacenamiento.py
import sys
import os
sys.path.append(os.path.dirname(__file__)+ "/../gestorAplicacion/modelos")
sys.path.append(os.path.dirname(__file__)+ "/../gestorAplicacion/reservacion")
sys.path.append(os.path.dirname(__file__)+ "/../uiMain/Exepciones")

import pickle
from Usuario import *
from Lugar import *
from Empleado import *
from Reserva import *
from Evento import *
from Servicio import *
from Factura import *
from ServicioExterno import *
from ExepcionC1 import *

class Almacenamiento():

    #Inicializar las listas para almacenado
    listaUsuarios = []
    listaEmpleados = []
    listaHabitaciones = []
    listaLugares = []
    listaReservas = []
    listaEventos = []
    listaServicios = []
    listaFacturas = []
    listaServiciosExternos = []
    listaHabitacionesDisponibles = []


    #Metodos de serializar y deserializar
    @classmethod
    def Serializar(cls):
        pathUsuarios = os.path.join(os.path.dirname(__file__), "temp/usuarios.txt")
        fileUsuarios=open(pathUsuarios , "wb")
        pickle.dump(Almacenamiento.listaUsuarios, fileUsuarios)
        fileUsuarios.close()

        pathEmpleados = os.path.join(os.path.dirname(__file__), "temp/empleados.txt")
        fileEmpleados = open(pathEmpleados, "wb")
        pickle.dump(Almacenamiento.listaEmpleados, fileEmpleados)
        fileEmpleados.close()

        pathHabitaciones = os.path.join(os.path.dirname(__file__), "temp/habitaciones.txt")
        fileHabitaciones = open(pathHabitaciones, "wb")
        pickle.dump(Almacenamiento.listaHabitaciones, fileHabitaciones)
        fileHabitaciones.close()

        pathHabitacionesDisponibles = os.path.join(os.path.dirname(__file__), "temp/habitacionesDisponibles.txt")
        fileHabitacionesDisponibles = open(pathHabitacionesDisponibles, "wb")
        pickle.dump(Almacenamiento.listaHabitacionesDisponibles, fileHabitacionesDisponibles)
        fileHabitacionesDisponibles.close()

        pathLugares = os.path.join(os.path.dirname(__file__), "temp/lugares.txt")
        fileLugares = open(pathLugares, "wb")
        pickle.dump(Almacenamiento.listaLugares, fileLugares)
        fileLugares.close()

        pathReservas = os.path.join(os.path.dirname(__file__), "temp/reservas.txt")
        fileReservas = open(pathReservas, "wb")
        pickle.dump(Almacenamiento.listaReservas, fileReservas)
        fileReservas.close()

        pathEventos = os.path.join(os.path.dirname(__file__), "temp/eventos.txt")
        fileEventos = open(pathEventos, "wb")
        pickle.dump(Almacenamiento.listaEventos, fileEventos)
        fileEventos.close()

        pathServicios = os.path.join(os.path.dirname(__file__), "temp/servicios.txt")
        fileServicios = open(pathServicios, "wb")
        pickle.dump(Almacenamiento.listaServicios, fileServicios)
        fileServicios.close()

        pathFacturas = os.path.join(os.path.dirname(__file__), "temp/facturas.txt")
        fileFacturas = open(pathFacturas, "wb")
        pickle.dump(Almacenamiento.listaFacturas, fileFacturas)
        fileFacturas.close()

        pathServiciosExternos = os.path.join(os.path.dirname(__file__), "temp/serviciosExternos.txt")
        fileServiciosExternos = open(pathServiciosExternos, "wb")
        pickle.dump(Almacenamiento.listaServiciosExternos, fileServiciosExternos)
        fileServiciosExternos.close()

        pathcontador = os.path.join(os.path.dirname(__file__), "temp/contadorFacturas.txt")
        filecontador = open(pathcontador, "wb")
        pickle.dump(Factura.contador, filecontador)
        filecontador.close()








    
    @classmethod
    def Deserializar(cls):
        pathUsuarios = os.path.join(os.path.dirname(__file__), "temp/usuarios.txt")
        fileUsuarios=open(pathUsuarios , "rb")
        Almacenamiento.listaUsuarios=pickle.load(fileUsuarios)
        fileUsuarios.close()

        pathEmpleados = os.path.join(os.path.dirname(__file__), "temp/empleados.txt")
        fileEmpleados = open(pathEmpleados, "rb")
        Almacenamiento.listaEmpleados = pickle.load(fileEmpleados)
        fileEmpleados.close()

        pathHabitaciones = os.path.join(os.path.dirname(__file__), "temp/habitaciones.txt")
        fileHabitaciones = open(pathHabitaciones, "rb")
        Almacenamiento.listaHabitaciones = pickle.load(fileHabitaciones)
        fileHabitaciones.close()

        pathHabitacionesDisponibles = os.path.join(os.path.dirname(__file__), "temp/habitacionesDisponibles.txt")
        fileHabitacionesDisponibles = open(pathHabitacionesDisponibles, "rb")
        Almacenamiento.listaHabitacionesDisponibles = pickle.load(fileHabitacionesDisponibles)
        fileHabitacionesDisponibles.close()

        pathLugares = os.path.join(os.path.dirname(__file__), "temp/lugares.txt")
        fileLugares = open(pathLugares, "rb")
        Almacenamiento.listaLugares = pickle.load(fileLugares)
        fileLugares.close()

        pathReservas = os.path.join(os.path.dirname(__file__), "temp/reservas.txt")
        fileReservas = open(pathReservas, "rb")
        Almacenamiento.listaReservas = pickle.load(fileReservas)
        fileReservas.close()

        pathEventos = os.path.join(os.path.dirname(__file__), "temp/eventos.txt")
        fileEventos = open(pathEventos, "rb")
        Almacenamiento.listaEventos = pickle.load(fileEventos)
        fileEventos.close()

        pathServicios = os.path.join(os.path.dirname(__file__), "temp/servicios.txt")
        fileServicios = open(pathServicios, "rb")
        Almacenamiento.listaServicios = pickle.load(fileServicios)
        fileServicios.close()

        pathFacturas = os.path.join(os.path.dirname(__file__), "temp/facturas.txt")
        fileFacturas = open(pathFacturas, "rb")
        Almacenamiento.listaFacturas = pickle.load(fileFacturas)
        fileFacturas.close()

        pathServiciosExternos = os.path.join(os.path.dirname(__file__), "temp/serviciosExternos.txt")
        fileServiciosExternos = open(pathServiciosExternos, "rb")
        Almacenamiento.listaServiciosExternos = pickle.load(fileServiciosExternos)
        fileServiciosExternos.close()

        pathcontador = os.path.join(os.path.dirname(__file__), "temp/contadorFacturas.txt")
        filecontador = open(pathcontador, "rb")
        Factura.contador= pickle.load(filecontador)
        filecontador.close()




    def crearUsuario(nombre, identificacion:int, telefono, cuentaBancaria):
        cliente=Usuario(nombre,identificacion,telefono,"Cliente",cuentaBancaria)
        Almacenamiento.listaUsuarios.append(cliente)

    def buscarUsuario(id:int) -> Usuario:
        usuario=None
        for cli in Almacenamiento.listaUsuarios:
            if (id==int(cli.getIdentificacion())):
                usuario=cli
                break
        return usuario         

    def crearHabitacion(nombre:str,descripcion,numero,capacidad):
        valor=Lugar.valorSegunTipo(descripcion)
        habitacion=Lugar(nombre,descripcion,valor,numero,capacidad)
        Almacenamiento.listaHabitaciones.append(habitacion)
        Almacenamiento.listaHabitacionesDisponibles.append(habitacion)

    def buscarHabitacion(numero) -> Lugar:
        habitacion=None
        for hab in Almacenamiento.listaHabitaciones:
            if (numero==hab.getNumero()):
                habitacion=hab
                break
        return habitacion
    
    def crearLugar(nombre:str,descripcion,numero,capacidad):
        valor=Lugar.valorSegunTipo(descripcion)
        lugar=Lugar(nombre,descripcion,valor,numero,capacidad)
        Almacenamiento.listaLugares.append(lugar)

    def buscarLugar(numero) -> Lugar:
        lugar=None
        for lug in Almacenamiento.listaLugares:
            if (numero==lug.getNumero()):
                lugar=lug
                break
        return lugar
    
    def crearEmpleado(nombre: str, identificacion, telefono, cargo):
        empleado=Empleado(nombre,identificacion,telefono,cargo)
        Almacenamiento.listaEmpleados.append(empleado)

    def buscarEmpleado(id) -> Empleado:
        Empleado=None
        for emp in Almacenamiento.listaEmpleados:
            if (id==emp.getIdentificacion()):
                Empleado=emp
                break
        return Empleado
    
    def crearReserva(fentrada,fsalida,habitaciones:list,aporte,cliente:Usuario):
        reserva=Reserva(fentrada,fsalida,habitaciones,aporte,cliente)
        Almacenamiento.listaReservas.append(reserva)
        return(reserva.__str__())

    def buscarReserva(cliente:Usuario) -> Reserva:
        Reserva=None
        for res in Almacenamiento.listaReservas:
            if (cliente==res.getCliente()):
                Reserva=res
                break
        return Reserva
    
    def crearEvento(codigo,lugar, cliente: Usuario, servicios:list, fecha, duracion, numeroAsistentes, empleados:Empleado,descripcion):
        evento=Evento(codigo,lugar,cliente,servicios,fecha,duracion,numeroAsistentes,empleados,descripcion)
        Almacenamiento.listaEventos.append(evento)

    def buscarEvento(cliente) -> Evento:
        Evento=None
        for even in Almacenamiento.listaEventos:
            if (cliente==even.getCliente()):
                Evento=even
                break
        return Evento
    
    def crearServicio(nombre, valor, descripcion):
        servicio=Servicio(nombre,valor,descripcion)
        Almacenamiento.listaServicios.append(servicio)

    def buscarServicio(nombre) -> Servicio:
        Servicio=None
        for ser in Almacenamiento.listaServicios:
            if (nombre==ser.getNombre()):
                Servicio=ser
                break
        return Servicio
    
    def crearServicioExterno(nombre, cliente: Usuario, descripcion):
        sExterno=ServicioExterno(nombre,cliente,descripcion)
        Almacenamiento.listaServiciosExternos.append(sExterno)

    def BuscarServicioExterno(cliente:Usuario) -> ServicioExterno:
        sExterno=None
        for ser in Almacenamiento.listaServiciosExternos:
            if (cliente==ser.getCliente()):
                sExterno=ser
                break
        return sExterno
    
    def crearFactura(cliente:Usuario, empleado: Empleado, listaItems: list, concepto):
        factura=Factura(cliente,empleado,listaItems,concepto)
        Almacenamiento.listaFacturas.append(factura)
        return (factura.imprimirFactura())
    
    def buscarFactura(cliente: Usuario) -> Factura:
        factura=None
        for fac in Almacenamiento.listaFacturas():
            if (cliente==fac.getUsuario()):
                factura=fac
                break
        return factura
    




